File: matrix_mult.py
import sys
import time

def main():
    if len(sys.argv) < 2:
        print("Uso: python matrix_mult.py <tamaño_N>", file=sys.stderr)
        sys.exit(1)

    N = int(sys.argv[1])
    if N <= 0:
        print("El tamaño N debe ser un entero positivo.", file=sys.stderr)
        sys.exit(1)

    A = [[float(i + j) for j in range(N)] for i in range(N)]
    B = [[float(i - j) for j in range(N)] for i in range(N)]
    C = [[0.0 for _ in range(N)] for _ in range(N)]

    start_time = time.perf_counter()

    for i in range(N):
        for j in range(N):
            total = 0.0
            for k in range(N):
                total += A[i][k] * B[k][j]
            C[i][j] = total

    end_time = time.perf_counter()

    elapsed_time = end_time - start_time
    print(f"{elapsed_time:.8f}")

if __name__ == "__main__":
    main()

File: matrix_mult.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Uso: %s <tamaño_N>\n", argv[0]);
        return 1;
    }

    int N = atoi(argv[1]);
    if (N <= 0) {
        fprintf(stderr, "El tamaño N debe ser un entero positivo.\n");
        return 1;
    }

    double *A = (double *)malloc((size_t)N * N * sizeof(double));
    double *B = (double *)malloc((size_t)N * N * sizeof(double));
    double *C = (double *)malloc((size_t)N * N * sizeof(double));

    if (!A || !B || !C) {
        fprintf(stderr, "Error de asignacion de memoria.\n");
        return 1;
    }

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            A[i * N + j] = (double)(i + j);
            B[i * N + j] = (double)(i - j);
            C[i * N + j] = 0.0;
        }
    }

    clock_t start = clock();

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            double sum = 0.0;
            for (int k = 0; k < N; k++) {
                sum += A[i * N + k] * B[k * N + j];
            }
            C[i * N + j] = sum;
        }
    }

    clock_t end = clock();

    double elapsed_time = (double)(end - start) / CLOCKS_PER_SEC;

    printf("%.8f\n", elapsed_time);

    free(A);
    free(B);
    free(C);

    return 0;
}
